java/code/src/com/redhat/rhn/frontend/xmlrpc/recurringaction/RecurringActionHandler.java

```diff
@@ -0,0 +1,243 @@
+import com.redhat.rhn.frontend.xmlrpc.TaskomaticApiException;
+import com.redhat.rhn.frontend.xmlrpc.ValidationException;
+import com.redhat.rhn.manager.recurringactions.RecurringActionManager;
+
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+public class RecurringActionHandler extends BaseHandler {
+
+    /* helper method */
+    private RecurringAction.Type getEntityType(String entityType) {
+        try {
+            return RecurringAction.Type.valueOf(entityType.toUpperCase());
+        }
+        catch (IllegalArgumentException e) {
+            throw new InvalidArgsException("Type \"" + entityType + "\" does not exist");
+        }
+    }
+
+    /**
+     * Return a list of recurring actions for a given entity.
+     *
+     * @param loggedInUser The current user
+     * @param entityId the id of the entity
+     * @param entityType type of the entity
+     * @return the list of recurring actions
+     *
+     * @xmlrpc.doc Return a list of recurring actions for a given entity.
+     * @xmlrpc.param #session_key()
+     * @xmlrpc.param #param_desc("string", "entityType", "Type of the target entity")
+     * @xmlrpc.param #param_desc("int", "entityId", "Id of the target entity")
+     * @xmlrpc.returntype
+     *      #array()
+     *          $RecurringActionSerializer
+     *      #array_end()
+     */
+    public List<? extends RecurringAction> listByEntity(User loggedInUser, String entityType, Integer entityId) {
+        List<? extends RecurringAction> schedules;
+        try {
+            switch (getEntityType(entityType)) {
+                case MINION:
+                    schedules = RecurringActionManager.listMinionRecurringActions(entityId, loggedInUser);
+                    break;
+                case GROUP:
+                    schedules = RecurringActionManager.listGroupRecurringActions(entityId, loggedInUser);
+                    break;
+                case ORG:
+                    schedules = RecurringActionManager.listOrgRecurringActions(entityId, loggedInUser);
+                    break;
+                default:
+                    throw new IllegalStateException("Unsupported type " + entityType);
+            }
+            return schedules;
+        }
+        catch (PermissionException e) {
+            throw new PermissionCheckFailureException(e.getMessage());
+        }
+    }
+
+    /**
+     * Return recurring action with given action id.
+     *
+     * @param loggedInUser The current user
+     * @param actionId id of the action
+     * @return the list of recurring actions
+     *
+     * @xmlrpc.doc Return recurring action with given action id.
+     * @xmlrpc.param #session_key()
+     * @xmlrpc.param #param_desc("int", "action_id", "Id of the action")
+     * @xmlrpc.returntype $RecurringActionSerializer
+     */
+    public RecurringAction lookupById(User loggedInUser, Integer actionId) {
+        try {
```

Comment:
This looks like improper use of `Optional` (accessing it and checking for `NoSuchElementException`).
Instead of that, something like this could replace the whole method body:
```java
RecurringActionFactory.lookupById(actionId)
    .filter(a -> a.canAccess(loggedInUser))
    .orElseThrow(() -> new EntityNotExistsFaultException("Action blabla doesnt exist or is not accessible"));

```